### PR TITLE
test: add calcTest for calc/calcout records (expected-behavior test)

### PR DESCRIPTION
**Pull request by jerzyjamroz**
# expected-behavior test

Covers:
- broken input link

This PR adds the missing test coverage for the calc and calcout records. Please review, if everything looks good, I will extend the tests further.

### CODE REVIEW

**Comment by mdavidsaver** [comment]
Right, I should have said more (https://github.com/epics-base/epics-base/issues/823#issuecomment-3972559000).  While calc and calcout records lack test code, `epicsCalcTest.cpp` covers the CALC parser and runtime.  So calc/calcout specific testing should focus on the record support code (`calcRecord.c` and `calcoutRecord.c`).  eg. use of `calcout` with `OOPT` to achieve conditional record processing.

**Comment by jerzyjamroz** [comment]
@mdavidsaver , I did the modification and focused just on the `test_calc_broken_inp_link();`

**Comment by jerzyjamroz** [comment]
So this test fails, and it is intended to present the expected behaviour:
```
# Set the record configuration
srcA.VAL DBF_DOUBLE:         10        
srcB.VAL DBF_DOUBLE:         12        
calc.INPA DBF_STRING:         "srcA NPP NMS"      
calc.INPB DBF_STRING:         "srcB NPP NMS"      
calc.SCAN DBF_STRING:         "Passive" 
# calc: normal calculation
calc.CALC DBF_STRING:         "A+B"     
calc.PROC DBF_UCHAR:          1 = 0x1   
ok  1 - calc.VAL: actual=22 expected=22
ok  2 - dbGetField("calc.SEVR", DBR_STRING) -> "NO_ALARM" == "NO_ALARM"
ok  3 - dbGetField("calc.STAT", DBR_STRING) -> "NO_ALARM" == "NO_ALARM"
ok  4 - dbGetField("calc.UDF", DBR_UCHAR) -> 0 == 0
# calc: break calc.INPB
calc.INPB DBF_STRING:         "__no_pv__ NPP NMS" 
# calc: broken INPB but expression uses only A
calc.CALC DBF_STRING:         "A"       
calc.PROC DBF_UCHAR:          1 = 0x1   
not ok  5 - calc.VAL: actual=22 expected=10
not ok  6 - dbGetField("calc.SEVR", DBR_STRING) -> "NO_ALARM" == "INVALID"
not ok  7 - dbGetField("calc.STAT", DBR_STRING) -> "NO_ALARM" == "LINK"
ok  8 - dbGetField("calc.UDF", DBR_UCHAR) -> 0 == 0
# calc: broken INPB but ternary selects A
calc.CALC DBF_STRING:         "1>0?A:B" 
calc.PROC DBF_UCHAR:          1 = 0x1   
not ok  9 - calc.VAL: actual=22 expected=10
not ok 10 - dbGetField("calc.SEVR", DBR_STRING) -> "NO_ALARM" == "INVALID"
not ok 11 - dbGetField("calc.STAT", DBR_STRING) -> "NO_ALARM" == "LINK"
ok 12 - dbGetField("calc.UDF", DBR_UCHAR) -> 0 == 0
# calc: broken INPB and expression requires B
calc.CALC DBF_STRING:         "A+B"     
calc.PROC DBF_UCHAR:          1 = 0x1   
not ok 13 - calc.VAL: actual=22 expected=NaN
ok 14 - dbGetField("calc.SEVR", DBR_STRING) -> "INVALID" == "INVALID"
ok 15 - dbGetField("calc.STAT", DBR_STRING) -> "LINK" == "LINK"
not ok 16 - dbGetField("calc.UDF", DBR_UCHAR) -> 1 == 0
# calc: broken INPB and expression is B only
calc.CALC DBF_STRING:         "B"       
calc.PROC DBF_UCHAR:          1 = 0x1   
not ok 17 - calc.VAL: actual=22 expected=NaN
ok 18 - dbGetField("calc.SEVR", DBR_STRING) -> "INVALID" == "INVALID"
ok 19 - dbGetField("calc.STAT", DBR_STRING) -> "LINK" == "LINK"
not ok 20 - dbGetField("calc.UDF", DBR_UCHAR) -> 1 == 0
```

**Comment by anjohnson** [comment]
Core Group: Please use the dbUnitTest facilities instead of dbpf(), and testDiag instead of printf().

The record change to not error out on disconnected links is interesting and worth doing.

**Comment by jerzyjamroz** [comment]
@mdavidsaver , For these link behaviour tests, is it enough to check only these fields: VAL, SEVR, STAT, and UDF? Or would you recommend adding any other fields?

This is what I currently use:
```
static void calc_expect_valid(double val)
{
    testDiag("CALC Expect Valid");
    testdbGetFieldEqual("calc.VAL", DBF_DOUBLE, (double)val);
    testdbGetFieldEqual("calc.SEVR", DBF_STRING, "NO_ALARM");
    testdbGetFieldEqual("calc.STAT", DBF_STRING, "NO_ALARM");
    testdbGetFieldEqual("calc.UDF",  DBF_USHORT,  0);
}

static void calc_expect_invalid(void)
{
    testDiag("CALC Expect Invalid");
    testdbGetFieldEqual("calc.SEVR", DBF_STRING, "INVALID");
    testdbGetFieldEqual("calc.STAT", DBF_STRING, "LINK");
    testdbGetFieldEqual("calc.UDF",  DBF_USHORT,  1);
}
```

Questions:

1. Should UDF become 1 in the invalid/link case, or is that not something we should rely on? (Should I remove it from this test case?).

2. If the calc record is SCAN="Passive", do you expect the record to be processed automatically when writing a new CALC expression, or should the test always force processing via PROC=1 to be deterministic? (I can add this test case).

3. When the record goes INVALID/LINK, should VAL be set to NaN, or is it intentionally unspecified (e.g. keeps the previous value)? (I can add this test case).

P.S.
In the next phase I plan to add similar coverage for the calcout record.

**Comment by jerzyjamroz** [comment]
This is ready for the review, but I think if the "calc" code is not corrected, it will always fail.

**Comment by jerzyjamroz** [comment]
I modified the test just to focus on this LINK issue.

File: modules/database/test/std/rec/calcTest.cpp [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    testdbPutFieldOk(pv, DBF_DOUBLE, (double)v);
+    testdbGetFieldEqual(pv, DBF_DOUBLE, (double)v);
+}
+static void testPutOk(const char *pv, const char *v)
+{
+    testDiag("put %s (string) = %s", pv, v);
+    testdbPutFieldOk(pv, DBF_STRING, v);
+    testdbGetFieldEqual(pv, DBF_STRING, v);
+}
+static void testPutOk(const char *pv, int v)
```
Comment (dirk-zimoch) [inline modules/database/test/std/rec/calcTest.cpp]: unused static function (makes -Werror builds fail)

Comment (mdavidsaver) [reply modules/database/test/std/rec/calcTest.cpp]: Also, the RTEMS test runner requires a correct non-zero `testPlan()` count.

**Comment by jerzyjamroz** [comment]
Ok, @mdavidsaver and @dirk-zimoch - in summary:
1) This PR starts behavioural tests for the calc/calcout records within the file calcTest.cpp
2) The first test documents how calc handles broken input links.

So the idea is to accept those two points and it will close this PR (which extends the issue https://github.com/epics-base/epics-base/issues/823). Other behavioural tests can be added within new PR-s in the future.

**Comment by jerzyjamroz** [comment]
The result looks like:
```
############################################################################
## EPICS R7.0.10.1-DEV
## Rev. R7.0.10-103-g09b01d2a81c7fd782cc8-dirty
## Rev. Date Git: 2026-06-11 11:04:04 +0200
############################################################################
# put srcA.VAL (double) = 10
ok  1 - dbPutField("srcA.VAL", DBR_DOUBLE, ...) -> 0 (Ok)
ok  2 - dbGetField("srcA.VAL", DBR_DOUBLE) -> 10 == 10
# put srcB.VAL (double) = 12
ok  3 - dbPutField("srcB.VAL", DBR_DOUBLE, ...) -> 0 (Ok)
ok  4 - dbGetField("srcB.VAL", DBR_DOUBLE) -> 12 == 12
# TEST sutCalc broken input link
# put sutCalc.INPA (string) = srcA.VAL NPP NMS
ok  5 - dbPutField("sutCalc.INPA", DBR_STRING, ...) -> 0 (Ok)
ok  6 - dbGetField("sutCalc.INPA", DBR_STRING) -> "srcA.VAL NPP NMS" == "srcA.VAL NPP NMS"
# put sutCalc.INPB (string) = srcB.VAL NPP NMS
ok  7 - dbPutField("sutCalc.INPB", DBR_STRING, ...) -> 0 (Ok)
ok  8 - dbGetField("sutCalc.INPB", DBR_STRING) -> "srcB.VAL NPP NMS" == "srcB.VAL NPP NMS"
# put sutCalc.SCAN (string) = Passive
ok  9 - dbPutField("sutCalc.SCAN", DBR_STRING, ...) -> 0 (Ok)
ok 10 - dbGetField("sutCalc.SCAN", DBR_STRING) -> "Passive" == "Passive"
# CASE sutCalc: normal calculation
# put sutCalc.CALC (string) = A+B
ok 11 - dbPutField("sutCalc.CALC", DBR_STRING, ...) -> 0 (Ok)
ok 12 - dbGetField("sutCalc.CALC", DBR_STRING) -> "A+B" == "A+B"
# process sutCalc
ok 13 - dbPutField("sutCalc.PROC", DBR_LONG, ...) -> 0 (Ok)
# sutCalc expect valid
ok 14 - dbGetField("sutCalc.VAL", DBR_DOUBLE) -> 22 == 22
ok 15 - dbGetField("sutCalc.SEVR", DBR_STRING) -> "NO_ALARM" == "NO_ALARM"
ok 16 - dbGetField("sutCalc.STAT", DBR_STRING) -> "NO_ALARM" == "NO_ALARM"
ok 17 - dbGetField("sutCalc.AMSG", DBR_STRING) -> "" == ""
ok 18 - dbGetField("sutCalc.NAMSG", DBR_STRING) -> "" == ""
ok 19 - dbGetField("sutCalc.UDF", DBR_USHORT) -> 0 == 0
# CASE sutCalc: break INPB
# put sutCalc.INPB (string) = __no_pv__ NPP NMS
ok 20 - dbPutField("sutCalc.INPB", DBR_STRING, ...) -> 0 (Ok)
ok 21 - dbGetField("sutCalc.INPB", DBR_STRING) -> "__no_pv__ NPP NMS" == "__no_pv__ NPP NMS"
# CASE sutCalc: broken INPB but expression uses only A
# put sutCalc.CALC (string) = A
ok 22 - dbPutField("sutCalc.CALC", DBR_STRING, ...) -> 0 (Ok) # TODO Broken LINK bug: unused broken input should not invalidate result
ok 23 - dbGetField("sutCalc.CALC", DBR_STRING) -> "A" == "A" # TODO Broken LINK bug: unused broken input should not invalidate result
# process sutCalc
ok 24 - dbPutField("sutCalc.PROC", DBR_LONG, ...) -> 0 (Ok) # TODO Broken LINK bug: unused broken input should not invalidate result
# sutCalc expect valid
not ok 25 - dbGetField("sutCalc.VAL", DBR_DOUBLE) -> 10 == 22 # TODO Broken LINK bug: unused broken input should not invalidate result
not ok 26 - dbGetField("sutCalc.SEVR", DBR_STRING) -> "NO_ALARM" == "INVALID" # TODO Broken LINK bug: unused broken input should not invalidate result
not ok 27 - dbGetField("sutCalc.STAT", DBR_STRING) -> "NO_ALARM" == "LINK" # TODO Broken LINK bug: unused broken input should not invalidate result
not ok 28 - dbGetField("sutCalc.AMSG", DBR_STRING) -> "" == "field INPB" # TODO Broken LINK bug: unused broken input should not invalidate result
not ok 29 - dbGetField("sutCalc.NAMSG", DBR_STRING) -> "" == "field INPB" # TODO Broken LINK bug: unused broken input should not invalidate result
ok 30 - dbGetField("sutCalc.UDF", DBR_USHORT) -> 0 == 0 # TODO Broken LINK bug: unused broken input should not invalidate result
# CASE sutCalc: broken INPB but ternary selects A
# put sutCalc.CALC (string) = 1>0?A:B
ok 31 - dbPutField("sutCalc.CALC", DBR_STRING, ...) -> 0 (Ok) # TODO Broken LINK bug: unselected ternary branch should not invalidate result
ok 32 - dbGetField("sutCalc.CALC", DBR_STRING) -> "1>0?A:B" == "1>0?A:B" # TODO Broken LINK bug: unselected ternary branch should not invalidate result
# process sutCalc
ok 33 - dbPutField("sutCalc.PROC", DBR_LONG, ...) -> 0 (Ok) # TODO Broken LINK bug: unselected ternary branch should not invalidate result
# sutCalc expect valid
not ok 34 - dbGetField("sutCalc.VAL", DBR_DOUBLE) -> 10 == 22 # TODO Broken LINK bug: unselected ternary branch should not invalidate result
not ok 35 - dbGetField("sutCalc.SEVR", DBR_STRING) -> "NO_ALARM" == "INVALID" # TODO Broken LINK bug: unselected ternary branch should not invalidate result
not ok 36 - dbGetField("sutCalc.STAT", DBR_STRING) -> "NO_ALARM" == "LINK" # TODO Broken LINK bug: unselected ternary branch should not invalidate result
not ok 37 - dbGetField("sutCalc.AMSG", DBR_STRING) -> "" == "field INPB" # TODO Broken LINK bug: unselected ternary branch should not invalidate result
not ok 38 - dbGetField("sutCalc.NAMSG", DBR_STRING) -> "" == "field INPB" # TODO Broken LINK bug: unselected ternary branch should not invalidate result
ok 39 - dbGetField("sutCalc.UDF", DBR_USHORT) -> 0 == 0 # TODO Broken LINK bug: unselected ternary branch should not invalidate result
# CASE sutCalc: broken INPB and expression requires B
# put sutCalc.CALC (string) = A+B
ok 40 - dbPutField("sutCalc.CALC", DBR_STRING, ...) -> 0 (Ok)
ok 41 - dbGetField("sutCalc.CALC", DBR_STRING) -> "A+B" == "A+B"
# process sutCalc
ok 42 - dbPutField("sutCalc.PROC", DBR_LONG, ...) -> 0 (Ok)
# sutCalc expect invalid
ok 43 - dbGetField("sutCalc.SEVR", DBR_STRING) -> "INVALID" == "INVALID"
ok 44 - dbGetField("sutCalc.STAT", DBR_STRING) -> "LINK" == "LINK"
ok 45 - dbGetField("sutCalc.AMSG", DBR_STRING) -> "field INPB" == "field INPB"
ok 46 - dbGetField("sutCalc.NAMSG", DBR_STRING) -> "field INPB" == "field INPB"
not ok 47 - dbGetField("sutCalc.UDF", DBR_USHORT) -> 1 == 0 # TODO UDF should be set for invalid LINK alarm
# CASE sutCalc: broken INPB and expression is B only
# put sutCalc.CALC (string) = B
ok 48 - dbPutField("sutCalc.CALC", DBR_STRING, ...) -> 0 (Ok)
ok 49 - dbGetField("sutCalc.CALC", DBR_STRING) -> "B" == "B"
# process sutCalc
ok 50 - dbPutField("sutCalc.PROC", DBR_LONG, ...) -> 0 (Ok)
# sutCalc expect invalid
ok 51 - dbGetField("sutCalc.SEVR", DBR_STRING) -> "INVALID" == "INVALID"
ok 52 - dbGetField("sutCalc.STAT", DBR_STRING) -> "LINK" == "LINK"
ok 53 - dbGetField("sutCalc.AMSG", DBR_STRING) -> "field INPB" == "field INPB"
ok 54 - dbGetField("sutCalc.NAMSG", DBR_STRING) -> "field INPB" == "field INPB"
not ok 55 - dbGetField("sutCalc.UDF", DBR_USHORT) -> 1 == 0 # TODO UDF should be set for invalid LINK alarm
# TEST sutCalcout broken input link
# put sutCalcout.INPA (string) = srcA.VAL NPP NMS
ok 56 - dbPutField("sutCalcout.INPA", DBR_STRING, ...) -> 0 (Ok)
ok 57 - dbGetField("sutCalcout.INPA", DBR_STRING) -> "srcA.VAL NPP NMS" == "srcA.VAL NPP NMS"
# put sutCalcout.INPB (string) = srcB.VAL NPP NMS
ok 58 - dbPutField("sutCalcout.INPB", DBR_STRING, ...) -> 0 (Ok)
ok 59 - dbGetField("sutCalcout.INPB", DBR_STRING) -> "srcB.VAL NPP NMS" == "srcB.VAL NPP NMS"
# put sutCalcout.SCAN (string) = Passive
ok 60 - dbPutField("sutCalcout.SCAN", DBR_STRING, ...) -> 0 (Ok)
ok 61 - dbGetField("sutCalcout.SCAN", DBR_STRING) -> "Passive" == "Passive"
# CASE sutCalcout: normal calculation
# put sutCalcout.CALC (string) = A+B
ok 62 - dbPutField("sutCalcout.CALC", DBR_STRING, ...) -> 0 (Ok)
ok 63 - dbGetField("sutCalcout.CALC", DBR_STRING) -> "A+B" == "A+B"
# process sutCalcout
ok 64 - dbPutField("sutCalcout.PROC", DBR_LONG, ...) -> 0 (Ok)
# sutCalcout expect valid
ok 65 - dbGetField("sutCalcout.VAL", DBR_DOUBLE) -> 22 == 22
ok 66 - dbGetField("sutCalcout.SEVR", DBR_STRING) -> "NO_ALARM" == "NO_ALARM"
ok 67 - dbGetField("sutCalcout.STAT", DBR_STRING) -> "NO_ALARM" == "NO_ALARM"
ok 68 - dbGetField("sutCalcout.AMSG", DBR_STRING) -> "" == ""
ok 69 - dbGetField("sutCalcout.NAMSG", DBR_STRING) -> "" == ""
ok 70 - dbGetField("sutCalcout.UDF", DBR_USHORT) -> 0 == 0
# CASE sutCalcout: break INPB
# put sutCalcout.INPB (string) = __no_pv__ NPP NMS
ok 71 - dbPutField("sutCalcout.INPB", DBR_STRING, ...) -> 0 (Ok)
ok 72 - dbGetField("sutCalcout.INPB", DBR_STRING) -> "__no_pv__ NPP NMS" == "__no_pv__ NPP NMS"
# CASE sutCalcout: broken INPB but expression uses only A
# put sutCalcout.CALC (string) = A
ok 73 - dbPutField("sutCalcout.CALC", DBR_STRING, ...) -> 0 (Ok) # TODO Broken LINK bug: unused broken input should not invalidate result
ok 74 - dbGetField("sutCalcout.CALC", DBR_STRING) -> "A" == "A" # TODO Broken LINK bug: unused broken input should not invalidate result
# process sutCalcout
ok 75 - dbPutField("sutCalcout.PROC", DBR_LONG, ...) -> 0 (Ok) # TODO Broken LINK bug: unused broken input should not invalidate result
# sutCalcout expect valid
not ok 76 - dbGetField("sutCalcout.VAL", DBR_DOUBLE) -> 10 == 22 # TODO Broken LINK bug: unused broken input should not invalidate result
not ok 77 - dbGetField("sutCalcout.SEVR", DBR_STRING) -> "NO_ALARM" == "INVALID" # TODO Broken LINK bug: unused broken input should not invalidate result
not ok 78 - dbGetField("sutCalcout.STAT", DBR_STRING) -> "NO_ALARM" == "LINK" # TODO Broken LINK bug: unused broken input should not invalidate result
not ok 79 - dbGetField("sutCalcout.AMSG", DBR_STRING) -> "" == "field INPB" # TODO Broken LINK bug: unused broken input should not invalidate result
not ok 80 - dbGetField("sutCalcout.NAMSG", DBR_STRING) -> "" == "field INPB" # TODO Broken LINK bug: unused broken input should not invalidate result
ok 81 - dbGetField("sutCalcout.UDF", DBR_USHORT) -> 0 == 0 # TODO Broken LINK bug: unused broken input should not invalidate result
# CASE sutCalcout: broken INPB but ternary selects A
# put sutCalcout.CALC (string) = 1>0?A:B
ok 82 - dbPutField("sutCalcout.CALC", DBR_STRING, ...) -> 0 (Ok) # TODO Broken LINK bug: unselected ternary branch should not invalidate result
ok 83 - dbGetField("sutCalcout.CALC", DBR_STRING) -> "1>0?A:B" == "1>0?A:B" # TODO Broken LINK bug: unselected ternary branch should not invalidate result
# process sutCalcout
ok 84 - dbPutField("sutCalcout.PROC", DBR_LONG, ...) -> 0 (Ok) # TODO Broken LINK bug: unselected ternary branch should not invalidate result
# sutCalcout expect valid
not ok 85 - dbGetField("sutCalcout.VAL", DBR_DOUBLE) -> 10 == 22 # TODO Broken LINK bug: unselected ternary branch should not invalidate result
not ok 86 - dbGetField("sutCalcout.SEVR", DBR_STRING) -> "NO_ALARM" == "INVALID" # TODO Broken LINK bug: unselected ternary branch should not invalidate result
not ok 87 - dbGetField("sutCalcout.STAT", DBR_STRING) -> "NO_ALARM" == "LINK" # TODO Broken LINK bug: unselected ternary branch should not invalidate result
not ok 88 - dbGetField("sutCalcout.AMSG", DBR_STRING) -> "" == "field INPB" # TODO Broken LINK bug: unselected ternary branch should not invalidate result
not ok 89 - dbGetField("sutCalcout.NAMSG", DBR_STRING) -> "" == "field INPB" # TODO Broken LINK bug: unselected ternary branch should not invalidate result
ok 90 - dbGetField("sutCalcout.UDF", DBR_USHORT) -> 0 == 0 # TODO Broken LINK bug: unselected ternary branch should not invalidate result
# CASE sutCalcout: broken INPB and expression requires B
# put sutCalcout.CALC (string) = A+B
ok 91 - dbPutField("sutCalcout.CALC", DBR_STRING, ...) -> 0 (Ok)
ok 92 - dbGetField("sutCalcout.CALC", DBR_STRING) -> "A+B" == "A+B"
# process sutCalcout
ok 93 - dbPutField("sutCalcout.PROC", DBR_LONG, ...) -> 0 (Ok)
# sutCalcout expect invalid
ok 94 - dbGetField("sutCalcout.SEVR", DBR_STRING) -> "INVALID" == "INVALID"
ok 95 - dbGetField("sutCalcout.STAT", DBR_STRING) -> "LINK" == "LINK"
ok 96 - dbGetField("sutCalcout.AMSG", DBR_STRING) -> "field INPB" == "field INPB"
ok 97 - dbGetField("sutCalcout.NAMSG", DBR_STRING) -> "field INPB" == "field INPB"
not ok 98 - dbGetField("sutCalcout.UDF", DBR_USHORT) -> 1 == 0 # TODO UDF should be set for invalid LINK alarm
# CASE sutCalcout: broken INPB and expression is B only
# put sutCalcout.CALC (string) = B
ok 99 - dbPutField("sutCalcout.CALC", DBR_STRING, ...) -> 0 (Ok)
ok 100 - dbGetField("sutCalcout.CALC", DBR_STRING) -> "B" == "B"
# process sutCalcout
ok 101 - dbPutField("sutCalcout.PROC", DBR_LONG, ...) -> 0 (Ok)
# sutCalcout expect invalid
ok 102 - dbGetField("sutCalcout.SEVR", DBR_STRING) -> "INVALID" == "INVALID"
ok 103 - dbGetField("sutCalcout.STAT", DBR_STRING) -> "LINK" == "LINK"
ok 104 - dbGetField("sutCalcout.AMSG", DBR_STRING) -> "field INPB" == "field INPB"
ok 105 - dbGetField("sutCalcout.NAMSG", DBR_STRING) -> "field INPB" == "field INPB"
not ok 106 - dbGetField("sutCalcout.UDF", DBR_USHORT) -> 1 == 0 # TODO UDF should be set for invalid LINK alarm

    Results
    =======
       Tests: 106
      Passed: 106 = 100.00%
 Todo Passes:  16 = 15.09%
```

**Comment by jerzyjamroz** [comment]
The test is adjusted according to the conclusion in: https://github.com/epics-base/epics-base/issues/823, so the broken link behaviour is kept the same as it was.
